File: packages-js/bpmn-flowaudit/src/model/legalBasis.ts
/**
 * Citation forms of a legal basis.
 *
 * The canonical form is the **long form** („Artikel 73 Absatz 2 Buchstabe b
 * der Verordnung (EU) 2021/1060“) because audit authority texts cite this
 * way (177 long to 59 short citations in the analysed diagrams). The short
 * form („Art. 73 Abs. 2 Buchst. b VO (EU) 2021/1060“) is for display.
 * Same rules as `Rechtsgrundlage.zitat()` in `auditcore_bpmn`.
 */

import type { LegalBasis } from '../schema/types'

const ACT_LONG: [RegExp, string][] = [
  [/^Delegierte\s+VO\b/, 'Delegierte Verordnung'],
  [/^Durchführungs-?VO\b/, 'Durchführungsverordnung'],
  [/^DVO\b/, 'Durchführungsverordnung'],
  [/^VO\b/, 'Verordnung'],
  [/^RL\b/, 'Richtlinie'],
]
const ACT_SHORT: [string, string][] = [
  ['Delegierte Verordnung', 'Delegierte VO'],
  ['Durchführungsverordnung', 'Durchführungs-VO'],
  ['Verordnung', 'VO'],
  ['Richtlinie', 'RL'],
]
const GENITIVE: [string, string][] = [
  ['Delegierte Verordnung', 'der Delegierten Verordnung'],
  ['Durchführungsverordnung', 'der Durchführungsverordnung'],
  ['Verordnung', 'der Verordnung'],
  ['Richtlinie', 'der Richtlinie'],
  ['Beschluss', 'des Beschlusses'],
  ['Entscheidung', 'der Entscheidung'],
]

/** Parts after the article, with short and long labels. */
const SUB_PARTS: [keyof LegalBasis, string, string][] = [
  ['paragraph', 'Abs.', 'Absatz'],
  ['subparagraph', 'UAbs.', 'Unterabsatz'],
  ['sentence', 'S.', 'Satz'],
  ['point', 'Buchst.', 'Buchstabe'],
  ['number', 'Nr.', 'Nummer'],
]

export const STRUCTURE_FIELDS = ['act', 'article', 'section', 'annex', 'paragraph', 'subparagraph', 'sentence', 'point', 'number'] as const

/** Long canonical form of an act name (`VO` → `Verordnung`). */
export function actLong(act: string): string {
  const value = act.split(/\s+/).filter(Boolean).join(' ')
  const rule = ACT_LONG.find(([pattern]) => pattern.test(value))
  return rule ? value.replace(rule[0], rule[1]) : value
}

/** Short form of an act name for display (`Verordnung` → `VO`). */
export function actShort(act: string): string {
  const value = actLong(act)
  const rule = ACT_SHORT.find(([long]) => value.startsWith(long))
  return rule ? rule[1] + value.slice(rule[0].length) : value
}

function actGenitive(act: string): string {
  const value = actLong(act)
  const rule = GENITIVE.find(([head]) => value === head || value.startsWith(`${head} `))
  return rule ? rule[1] + value.slice(rule[0].length) : value
}

export function isStructured(basis: LegalBasis): boolean {
  return STRUCTURE_FIELDS.some((field) => Boolean(basis[field]))
}

function head(basis: LegalBasis, short: boolean): string[] {
  if (basis.article) return [`${short ? 'Art.' : 'Artikel'} ${basis.article}`]
  if (basis.section) return [`§ ${basis.section}`]
  if (basis.annex) return [`Anhang ${basis.annex}`]
  return []
}

function parts(basis: LegalBasis, short: boolean): string[] {
  const tail = SUB_PARTS.filter(([field]) => basis[field]).map(([field, s, l]) => `${short ? s : l} ${String(basis[field])}`)
  return [...head(basis, short), ...tail]
}

/** `VV zu § 44 LHO` with number → „VV Nummer 4.2 zu § 44 LHO“. */
function administrativeRule(basis: LegalBasis, short: boolean): string | null {
  if (basis.act?.startsWith('VV zu ') && basis.number && !(basis.article || basis.section || basis.annex)) {
    return `VV ${short ? 'Nr.' : 'Nummer'} ${basis.number} ${basis.act.slice(3)}`
  }
  return null
}

/** Long citation, e.g. „Artikel 73 Absatz 2 Buchstabe b der Verordnung (EU) 2021/1060“. */
export function citation(basis: LegalBasis): string {
  if (!isStructured(basis)) return (basis.text ?? '').trim()
  const rule = administrativeRule(basis, false)
  if (rule) return rule
  const result = parts(basis, false)
  if (basis.act) {
    const genitive = Boolean(basis.article || basis.annex) && result.length > 0
    result.push(genitive ? actGenitive(basis.act) : actLong(basis.act))
  }
  return result.join(' ')
}

/** Short citation for display, e.g. „Art. 73 Abs. 2 Buchst. b VO (EU) 2021/1060“. */
export function shortCitation(basis: LegalBasis): string {
  if (!isStructured(basis)) return (basis.text ?? '').trim()
  const rule = administrativeRule(basis, true)
  if (rule) return rule
  const result = parts(basis, true)
  if (basis.act) result.push(actShort(basis.act))
  return result.join(' ')
}

/** Readable text: free text, otherwise short citation. */
export function displayText(basis: LegalBasis): string {
  return (basis.text ?? '').trim() || shortCitation(basis)
}

/** Act in long form and text content as long citation. */
export function normalized(basis: LegalBasis): LegalBasis {
  if (!isStructured(basis)) return basis
  const result: LegalBasis = { ...basis }
  if (basis.act) result.act = actLong(basis.act)
  result.text = citation(result)
  return result
}

/**
 * Prepares a legal basis for writing: text content is mandatory (1.0 readers
 * show it). Existing free text is never overwritten.
 */
export function forWriting(basis: LegalBasis): LegalBasis {
  if (basis.text?.trim()) return basis
  return isStructured(basis) ? { ...basis, text: citation(basis) } : basis
}

/** Splits a 1.0 free text with several acts („§ 55 BHO; Art. 74 VO (EU) 2021/1060“). */
export function splitFreeText(text: string): string[] {
  return text
    .split(/\s*[;\n]\s*/)
    .map((part) => part.trim())
    .filter(Boolean)
}

/** Key for duplicate detection (structured fields or text). */
export function legalBasisKey(basis: LegalBasis): string {
  if (!isStructured(basis)) return (basis.text ?? '').trim().toLowerCase().replace(/\s+/g, ' ')
  return STRUCTURE_FIELDS.map((field) => (field === 'act' && basis.act ? actLong(basis.act) : basis[field] ?? ''))
    .join('|')
    .toLowerCase()
}

File: packages/auditcore_bpmn/src/auditcore_bpmn/extensions/legal_basis.py
"""``flowaudit:rechtsgrundlage``: Freitext (Schema 1.0) und strukturierte Fundstelle (1.1).

Normalform ist die ausgeschriebene Fassung („Artikel 73 Absatz 2 Buchstabe b
der Verordnung (EU) 2021/1060“), die Kurzform („Art. 73 Abs. 2 Buchst. b
VO (EU) 2021/1060“) dient der Anzeige.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, replace

from .mapping import xml_field

_ACT_LONG = (
    (re.compile(r"^Delegierte\s+VO\b"), "Delegierte Verordnung"),
    (re.compile(r"^Durchführungs-?VO\b"), "Durchführungsverordnung"),
    (re.compile(r"^DVO\b"), "Durchführungsverordnung"),
    (re.compile(r"^VO\b"), "Verordnung"),
    (re.compile(r"^RL\b"), "Richtlinie"),
)
_ACT_SHORT = (
    ("Delegierte Verordnung", "Delegierte VO"),
    ("Durchführungsverordnung", "Durchführungs-VO"),
    ("Verordnung", "VO"),
    ("Richtlinie", "RL"),
)
_GENITIVE = {
    "Verordnung": "der Verordnung",
    "Delegierte Verordnung": "der Delegierten Verordnung",
    "Durchführungsverordnung": "der Durchführungsverordnung",
    "Richtlinie": "der Richtlinie",
    "Beschluss": "des Beschlusses",
    "Entscheidung": "der Entscheidung",
}
#: (Feld, Kurzform, Langform) der Untergliederung in Zitierreihenfolge.
_SUBDIVISIONS = (
    ("paragraph", "Abs.", "Absatz"),
    ("subparagraph", "UAbs.", "Unterabsatz"),
    ("sentence", "S.", "Satz"),
    ("point", "Buchst.", "Buchstabe"),
    ("number", "Nr.", "Nummer"),
)


def act_long(act: str) -> str:
    """Ausgeschriebener Normname (``VO`` → ``Verordnung``)."""
    value = " ".join(act.split())
    for pattern, replacement in _ACT_LONG:
        if pattern.search(value):
            return pattern.sub(replacement, value, count=1)
    return value


def act_short(act: str) -> str:
    """Kurzer Normname für die Anzeige (``Verordnung`` → ``VO``)."""
    value = act_long(act)
    for long, short in _ACT_SHORT:
        if value.startswith(long):
            return short + value[len(long) :]
    return value


def _act_with_article(act: str) -> str:
    value = act_long(act)
    for head, genitive in _GENITIVE.items():
        if value == head or value.startswith(head + " "):
            return genitive + value[len(head) :]
    return value


@dataclass(frozen=True)
class LegalBasis:
    """Rechtsgrundlage; ``article``/``section``/``annex`` enthalten nur die Kennung."""

    text: str | None = xml_field("", "body")
    id: str | None = xml_field("id")
    act: str | None = xml_field("norm")
    article: str | None = xml_field("artikel")
    section: str | None = xml_field("paragraph")
    annex: str | None = xml_field("anhang")
    paragraph: str | None = xml_field("absatz")
    subparagraph: str | None = xml_field("unterabsatz")
    sentence: str | None = xml_field("satz")
    point: str | None = xml_field("buchstabe")
    number: str | None = xml_field("nummer")
    version: str | None = xml_field("fassung")
    eli: str | None = xml_field("eli")
    celex: str | None = xml_field("celex")
    url: str | None = xml_field("url")
    short_title: str | None = xml_field("kurzbezeichnung")
    note: str | None = xml_field("anmerkung")
    confidential: bool | None = xml_field("vertraulich", bool_value=True)

    @property
    def is_structured(self) -> bool:
        """``True``, wenn mindestens ein Strukturfeld gesetzt ist."""
        return any(
            (
                self.act,
                self.article,
                self.section,
                self.annex,
                self.paragraph,
                self.subparagraph,
                self.sentence,
                self.point,
                self.number,
            )
        )

    def _administrative_rule(self, short: bool) -> str | None:
        """``VV zu § 44 LHO`` mit Nummer → „VV Nummer 4.2 zu § 44 LHO“."""
        if (
            self.act
            and self.act.startswith("VV zu ")
            and self.number
            and not (self.article or self.section or self.annex)
        ):
            return f"VV {'Nr.' if short else 'Nummer'} {self.number} {self.act[3:]}"
        return None

    def _parts(self, short: bool) -> list[str]:
        parts: list[str] = []
        if self.article:
            parts.append(f"{'Art.' if short else 'Artikel'} {self.article}")
        elif self.section:
            parts.append(f"§ {self.section}")
        elif self.annex:
            parts.append(f"Anhang {self.annex}")
        for name, short_label, long_label in _SUBDIVISIONS:
            value = getattr(self, name)
            if value:
                parts.append(f"{short_label if short else long_label} {value}")
        return parts

    def citation(self) -> str:
        """Ausgeschriebene Normalform; ohne Struktur der Freitext."""
        if not self.is_structured:
            return (self.text or "").strip()
        if rule := self._administrative_rule(short=False):
            return rule
        parts = self._parts(short=False)
        if self.act:
            linked = (self.article or self.annex) and parts
            parts.append(_act_with_article(self.act) if linked else act_long(self.act))
        return " ".join(parts)

    def short_citation(self) -> str:
        """Kurzform für die Anzeige."""
        if not self.is_structured:
            return (self.text or "").strip()
        if rule := self._administrative_rule(short=True):
            return rule
        parts = self._parts(short=True)
        if self.act:
            parts.append(act_short(self.act))
        return " ".join(parts)

    @property
    def display(self) -> str:
        """Lesbare Angabe: Freitext, sonst Kurzform."""
        return (self.text or "").strip() or self.short_citation()

    def normalized(self) -> LegalBasis:
        """Norm in Langform, Textinhalt = ausgeschriebene Normalform."""
        if not self.is_structured:
            return self
        result = replace(self, act=act_long(self.act) if self.act else None)
        return replace(result, text=result.citation())
